src/server/db_slice.cc

```diff
@@ -1410,4 +1413,87 @@ void DbSlice::TrackKeys(const facade::Connection::WeakRef& conn, const ArgSlice&
   }
 }
 
+void DbSlice::SendInvalidationTrackingMessage(std::string_view key) {
+  auto it = client_tracking_map_.find(key);
+  if (it != client_tracking_map_.end()) {
+    // notify all the clients.
+    auto& client_set = it->second;
+    auto cb = [key, client_set = std::move(client_set)](unsigned idx, util::ProactorBase*) {
+      for (auto it = client_set.begin(); it != client_set.end(); ++it) {
+        if ((unsigned int)it->Thread() != idx)
+          continue;
+        facade::Connection* conn = it->Get();
+        if ((conn != nullptr) && conn->IsTrackingOn()) {
+          std::string key_str = {key.begin(), key.end()};
+          conn->SendInvalidationMessageAsync({key_str});
+        }
+      }
+    };
+    shard_set->pool()->DispatchBrief(std::move(cb));
+    // remove this key from the tracking table as the key no longer exists
+    client_tracking_map_.erase(key);
+  }
+}
+
+void DbSlice::PerformDeletion(PrimeIterator del_it, ExpireIterator exp_it, EngineShard* shard,
+                              DbTable* table) {
+  std::string tmp;
+  std::string_view key = del_it->first.GetSlice(&tmp);
+
+  if (!exp_it.is_done()) {
+    table->expire.Erase(exp_it);
+  }
+
+  if (del_it->second.HasFlag()) {
+    if (table->mcflag.Erase(del_it->first) == 0) {
+      LOG(ERROR) << "Internal error, inconsistent state, mcflag should be present but not found "
+                 << del_it->first.ToString();
+    }
+  }
+
+  DbTableStats& stats = table->stats;
+  const PrimeValue& pv = del_it->second;
+  if (pv.IsExternal()) {
+    auto [offset, size] = pv.GetExternalSlice();
+
+    stats.tiered_entries--;
+    stats.tiered_size -= size;
+    TieredStorage* tiered = shard->tiered_storage();
+    tiered->Free(offset, size);
+  }
+  if (pv.HasIoPending()) {
+    TieredStorage* tiered = shard->tiered_storage();
+    tiered->CancelIo(table->index, del_it);
+  }
+
+  size_t value_heap_size = pv.MallocUsed();
+  stats.inline_keys -= del_it->first.IsInline();
+  int64_t delta = del_it->first.MallocUsed() + value_heap_size;
+  stats.obj_memory_usage -= delta;
+  stats.AddTypeMemoryUsage(pv.ObjType(), -delta);
+  if (pv.ObjType() == OBJ_HASH && pv.Encoding() == kEncodingListPack) {
+    --stats.listpack_blob_cnt;
+  } else if (pv.ObjType() == OBJ_ZSET && pv.Encoding() == OBJ_ENCODING_LISTPACK) {
+    --stats.listpack_blob_cnt;
+  }
+
+  if (ClusterConfig::IsEnabled()) {
+    SlotId sid = ClusterConfig::KeySlot(key);
+    table->slots_stats[sid].key_count -= 1;
+  }
+
+  table->prime.Erase(del_it);
+  SendInvalidationTrackingMessage(key);
+}
+
+void DbSlice::PerformDeletion(PrimeIterator del_it, EngineShard* shard, DbTable* table) {
```

Comment:
After reviewing this more times and reviewing other PR in this file I now understand that actually PerformDeletion should not be a member of DbSlice as I asked before. Rather we should have in DbSlice and function DeleteInternal which will write to journal, do the doc_del_cb_ call, SendInvalidationTrackingMessage
This function will be called from all the flows except the flush flows in which we dont need to do this changes and there we will call PerformDeletion which will be static function.

I believe this will fix several bugs we have that we forget to call the doc_del_cb_ and the journal change writes

Reply:
This kind of refactoring is out of the scope for this PR. We will do it in another one once this gets merged.

Reply:
Yes the refactoring can be done in a seprate PR. But as I wrote in another comment I believe that sending Invalidation message from PerformDeletion is incorrect for the flow of flush all.

Reply:
OK, please see if my reply above will resolve this issue as well.